src/query/api/v1/handler/prometheus/native/read.go

```diff
@@ -138,11 +138,12 @@ func (h *PromReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	result, params, respErr := h.ServeHTTPWithEngine(w, r, h.engine, &executor.EngineOptions{
-		QueryContextOptions: models.QueryContextOptions{
-			LimitMaxTimeseries: fetchOpts.Limit,
-		},
+	engineOpts := h.engine.Opts().SetQueryContextOptions(models.QueryContextOptions{
+		LimitMaxTimeseries: fetchOpts.Limit,
 	})
+	h.engine = h.engine.SetOpts(engineOpts)
```

Comment:
This feels really racy if multiple requests mutate `h.engine`; may be better to keep the QueryContextOptions separate

Reply:
+1 -- if this is per call, pass it down in the method args.

Reply:
Yeah, do not want to mutate the engine itself on a per call basis.